sql,pgwire: fixes for float formatting

rafiss opened the pull request:
### sql/pgwire: update float formatting to be compatible with PG12

Release note (sql change): The extra_float_digits session variable now
defaults to 1. The meaning of the variable has also changed. Now, any
value greater than zero causes floats to be formatted in their
shortest precise decimal representation. That is, the string representation
produced is closer to the actual binary value than to any other value.
(Previously, this was only the behavior when extra_float_digits was set
to 3.) This change was made in accordance with an equivalent change that
was part of the PostgreSQL 12.0 release.

The behavior of a non-positive extra_float_digits value is unchanged. That
will still reduce the number of float digits shown in the output string.
The formula for number of digits shown is `max(1, (DIGITS + extra_float_digits))`,
where DIGITS=6 for FLOAT4 values, and DIGITS=15 for FLOAT8 values.

### sql/pgwire: fix formatting of float arrays

Release note (bug fix): Fixed the formatting of floats in arrays sent
over the client-server pgwire protocol so that they respect the
extra_float_digits parameter, and correctly  format infinity values.

### eval: use common function for float->text cast

There's no behavior change, but this reduces code duplication.

This also adds a test that demonstrates a gap in our type/cast system.

rafiss committed:
sql/pgwire: update float formatting to be compatible with PG12

Release note (sql change): The extra_float_digits session variable now
defaults to 1. The meaning of the variable has also changed. Now, any
value greater than zero causes floats to be formatted in their
shortest precise decimal representation. That is, the string representation
produced is closer to the actual binary value than to any other value.
(Previously, this was only the behavior when extra_float_digits was set
to 3.) This change was made in accordance with an equivalent change that
was part of the PostgreSQL 12.0 release.

The behavior of a non-positive extra_float_digits value is unchanged. That
will still reduce the number of float digits shown in the output string.
The formula for number of digits shown is `max(1, (DIGITS + extra_float_digits))`,
where DIGITS=6 for FLOAT4 values, and DIGITS=15 for FLOAT8 values.

rafiss committed:
pgwire: fix formatting of float arrays and tuples

Release note (bug fix): Fixed the formatting of floats in arrays and tuples
sent over the client-server pgwire protocol so that they respect the
extra_float_digits parameter, and correctly  format infinity values.

rafiss committed:
eval: use common function for float->text cast

There's no behavior change, but this reduces code duplication.

This also adds a test that demonstrates a gap in our type/cast system.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/82022)
<!-- Reviewable:end -->

knz reviewed (commented):
Nice - but I'll let oliver sign off.

> Reviewed 6 of 6 files at r1, 3 of 3 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82022#-:-N3B7dFNAKo8XHdwT5k7:b2f78i8)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)



<!-- Sent from Reviewable.io -->

otan commented on pkg/sql/pgwire/testdata/pgtest/float:
```diff
@@ -0,0 +1,56 @@
+
+
+send
+Query {"String": "SET extra_float_digits = -1"}
```
can you add a test for a positive extra_float_digits > 1?

rafiss replied:
adding

otan commented on pkg/sql/pgwire/testdata/pgtest/float:
```diff
@@ -0,0 +1,56 @@
+ReadyForQuery
+----
+{"Type":"RowDescription","Fields":[{"Name":"float4","TableOID":0,"TableAttributeNumber":0,"DataTypeOID":700,"DataTypeSize":4,"TypeModifier":-1,"Format":0},{"Name":"float8","TableOID":0,"TableAttributeNumber":0,"DataTypeOID":701,"DataTypeSize":8,"TypeModifier":-1,"Format":0}]}
+{"Type":"DataRow","Values":[{"text":"0.33333334"},{"text":"0.3333333333333333"}]}
```
ooh we round up to 4? is this consistent when running this with PG?

rafiss replied:
yep,  the expected result was obtained by running against PG

otan commented on pkg/sql/sem/tree/pgwire_encode.go:
```diff
@@ -162,6 +170,27 @@ func init() {
 
+// PgwireFormatFloat returns a []byte representing a float according to
+// pgwire encoding. The result is appended to the given buffer.
+func PgwireFormatFloat(
```
wonder if this should be PGWire, but that's an optional nit

rafiss replied:
i was matching the naming of PgwireFormatBool

rafiss committed:
pgwire: update encoding tests to use new default

The encoding tests were updated so they use the new default value of 1
for extra_float_digits. This makes the test more aligned to Postgres
behavior.

Release note: None

rafiss commented:
bors r=otan

rafiss commented:
bors r=otan
